sys/net/application_layer/nanocoap/sock.c

```diff
@@ -46,19 +46,14 @@ ssize_t nanocoap_request(coap_pkt_t *pkt, sock_udp_ep_t *local, sock_udp_ep_t *r
 
     /* TODO: timeout random between between ACK_TIMEOUT and (ACK_TIMEOUT *
      * ACK_RANDOM_FACTOR) */
-    uint32_t timeout = COAP_ACK_TIMEOUT * (1000000U);
-    int tries = 0;
-    while (tries++ < COAP_MAX_RETRANSMIT) {
-        if (!tries) {
-            DEBUG("nanocoap: maximum retries reached.\n");
-            res = -ETIMEDOUT;
-            goto out;
-        }
+    uint32_t timeout = COAP_ACK_TIMEOUT * (1 * US_PER_SEC);
+    unsigned tries_left = COAP_MAX_RETRANSMIT + 1;  /* add 1 for initial transmit */
+    while (tries_left) {
```

Comment:
I guess the *correct* (more C'ish) (?) way would be to use `do { ... } while ( ... );`, like:

```
unsigned tries_left = COAP_MAX_RETRANSMIT;
do {
[...]
} while (tries_left--);
```

This should also work for `COAP_MAX_RETRANSMIT = 0`, that is: send once and no retries.
though I also guess this will not impact code size, as the compiler will optimise both the same way.

Reply:
In my experience, use of the decrement operator on an unsigned value is error prone. In this case, how do I display the "maximum retries reached" debug message when max retransmit is zero? I tested, and it doesn't work. The decrement operator rolled it back over.

I'm OK with converting tries_left to a signed value, but then I would want to use an explicit inequality in the do test "tries_left-- > 0". If I don't the logic would look strange against the test for debug below.

```
    } while (tries_left-- > 0);

    if (tries_left <= 0) {
        DEBUG("nanocoap: maximum retries reached\n");
    }
```

I still like this solution better than my original because I think the overall flow is easier to follow. Let me know what you think.

Reply:
yeah, agreed - leave as is